### ISSUE TITLE
Implement electric locomotives

### ISSUE THREAD

**Issue by cam72cam** [cam72cam/ImmersiveRailroading#21]
Feed from overhead wires from IE?  Does that chunkloading work?  Do we want them battery powered?

**Comment by skyem123** [comment]
I have no idea how (un)helpful this'll be, but I'll dump everything I know in here: In the UK, there are three (or five) ways to get power for electric trains:

* Third Rail - an extra rail on the outside of the running rails, powered at around 750V DC. if you step on it... zzt.
  * Fourth rail - as well as the third rail, a rail in between the running rails, either grounded or powered at -200V DC (if so the third rail is powered at +300V DC) - used on the London underground because the running rails weren't designed to handle current returning through them!
  * underside contact third rail - safer than third rail, you won't be electrocuted if you stand on it!

* overhead electrification - there are many different voltages, but in the UK and high speed rails across the world, it is 20kV 50Hz AC. The overhead lines are quite complex and have to be kept taut so the pantograph pressing up on them doesn't lift them! There tends to be a system of weights at certain points to hold the wires.

* Battery - Only really found in maintenance trains in the London Underground, useful for when there's no power, such as maintenance when the power is switched off. 😛

**Comment by WrongWeekend** [comment]
I've mentioned this before but I think it's worth putting here:

Battery powered electrics may be overpowered in survival, because in most tech mods (IE included), setting up an electricity supply + crafting energy storage items is done sooner than producing surplus diesel fuel. Combine this with the superior performance of electrics and diesels become obsolete before ever being an option.

**Comment by SuspiciousMango** [comment]
In regards to the battery idea:

I think its an ok solution for a little while just to get the engines into the mod under the pretense that it will be removed once a proper solution is implemented. If people want to be able to play around with electric locomotives before cam gets around to coding the third rail or overhead wire solutions, the battery option can take the existing inventory structure and modify it slightly to fit electric locomotives. 

This of course is only a reasonable solution if the demand for electric locomotives gets outrageous before cam decides to code what is required, but its something that I feel should be talked about a little bit.

**Comment by skyem123** [comment]
In real life battery electric trains have the limitation of being slow and heavy and also needing to recharge quite often. Maybe the in game limitation could be that it's a pain to use them?

**Comment by Trollebas** [comment]
NO, https://im0-tub-ru.yandex.net/i?id=98d61d2e9fcfdcdae8745990a23a717e-srl&n=13 вот провода на жд какие надо, пусть и русские, но не из Immersive Enginering

**Comment by txlawso** [comment]
Idea: Electrification is an option when building rails. A electrified third rail will do damage when touched, and requires extra materials, probably copper. Electrified overhead wires require LV or MV wire option for covered(No damage) or uncovered(Take damage when touched). All electric elements are powered by a rail charger which depends on the type of electrification ,the range it electrifies. Trains could have internal batteries for when there is no electrification(Electrified third rail doesn’t work with switches)

**Comment by CalocoDoesGit** [comment]
Electric trains don't really have onboard storage for traction power. Just to keep lights and various systems operating when temporarily overhead or third rail power is cut.
Trains who go past a section of unpowered track use the already gained momentum to continue on to the next section of electrified trackage.

**Comment by WrongWeekend** [comment]
The main obstacle to getting electric track and trains working properly in Immersive Railroading is developing an efficient system to transfer quantities of electric power through unloaded chunks. Such a system must be developed before any sort of track electrification- overhead wire, third rail, and so on- can be implemented.

**Comment by EpicPuppy613** [comment]
You could have the chunks stay loaded when a section of rail is electrified, like my idea before, when a train is within the range of a rail electrifier only the section that the electrifier is electrifying is loaded(Ex. There is a 16 chunk section of electrified track, only when the train is within those 16 chunks all 16 of them are loaded.)

**Comment by Xyvoracle** [comment]
That doesn't solve the issue of knowing which network said loaded track is actually a part of though, unfortunately. That would require the entire network to be loaded or having some kind of means of storing that information when it is initially created or when the network is updated through some other means than chunk loading.

**Comment by Cpt-Hellfire** [comment]
Wouldn't pre-generation of a world "fix" the problem or powering a certain amount of chunks at a time as they load in and out?

**Comment by Xyvoracle** [comment]
Define pre-generation, that sounds the same as chunk loading

**Comment by Cpt-Hellfire** [comment]
Its preloading chunks within a set square.

**Comment by FriedrichLP** [comment]
Loading chunks is definitely not the way to go since they eat up tons of memory and cpu resources. The closest way would be to assign an ID to each individual network and then globally store all ID's and their assigned power values. Then the trains could do a very cheap lookup and use that to also drain power. The complicated part is actually detecting those individual networks and then handling the cases where one gets split into two

**Comment by WrongWeekend** [comment]
There is also the potential issue of networks getting broken into three or more sections at once (such as if an electrified switch is destroyed, or multiple non-adjacent sections of electrified track are destroyed on the same tick) or multiple electrified rail networks getting merged together into one.

**Comment by WrongWeekend** [comment]
I believe the voltage and current type of the overhead wires and third rail should affect what types of locomotives can use them in a manner similar to track gauge. This kind of mechanic could be as simple as selecting a voltage/current type (such as 3000 V DC or 11kV 25 Hz AC) when placing the wires or wired track. Wire sections set to different voltages would not connect to each other. More importantly, electric locomotives designed for a certain type of electrification would not be able to run on a different one.

This has significant historical basis, as the type of electrification used on a railroad had a major impact on how electric locomotives were designed for it. AC electrification is more economical for longer distances, while DC is easier to set up over shorter distances. DC traction locomotives running from DC wires can be built to a simpler design. Locomotives running on AC current require rectifiers or heavy motor-generator setups to convert the AC to DC for DC traction motors, or use specialized AC motors. Since the types of motors used by electric locomotives are partially determined by the electrification system and have a major effect on the locomotive's performance, this means that the electrification system that an electric locomotive uses would play an important role in balancing its usefulness with other locomotives.

To give a real world example, most of the Milwaukee Road's locomotives, including the Little Joes, utilized 3000V DC, and most of the PRR fleet including the GG1s used 11kV 25 Hz AC. A locomotive built for the Milwaukee Road would be completely incapable of running on the PRR's electrification system and vice versa. By contrast, the Great Northern used the same 11kV 25 Hz AC electrification as the PRR, and a number of ex-GN electric locomotives were able to run on the PRR system without any major modifications.

**Comment by jcc10** [comment]
A alternative to chunk-loading would be to require a electrifier every so many chunks (so there is always one loaded) and leave the electric network stuff for whatever power mod is being used. A low capacity capacitor would probably be good to ensure that if you are moving through a chunk or two that is missing power the device remains active.

The same mechanics would be useful for cable cars such as the ones found in SF.

**Comment by Trollebas** [comment]
In general, after reading the description above, I think that the implementation of the overhead wire should also have a tension, because without it you can't accelerate on a overhead wire, and you also lose energy, and you can't put anything above the pantograph rope.

**Comment by TheGamineer** [comment]
Definitely a bit late to the discussion here, but what if we added a system that saves and maps the points between different ends of powered rail, and checks whether they were powered upon last being loaded in. Then we can take the trains locational information that already seems to exists, and see if it lines up between these electrified "points." It's just an idea, but it might be an avid fix to the chunk loading problem that seems to plague electric power.
![Screenshot 2024-09-06 101218](https://github.com/user-attachments/assets/17b95546-7330-4076-b857-2c176ee40781)

**Comment by Raspaholic** [comment]
Idk how much of use my comment will be, but from the last exchange & img I believe that might be the most simplest way to be able to really implement "electric" trains. Following the img, I think this also limits any lag or server load necessary to monitor/verify power delivery to the rail line across chunks. 

If there's still consideration for catenaries with electric lines, ya'll could just implement it as a cosmetic (optional) when building an electric line with the Track Blueprint 🤷‍♂ 

Either way, I believe that the img is the closest solution to implementing electric locos in IR while not putting strain on resources. Obviously it's up to ya'll IR devs to decide how to add them without rendering Diesel/Steam locos useless early-game (as mentioned by @WrongWeekend)
